### ISSUE TITLE
Zookeeper plugin: if value is not a integer it is casted to string. Average values are float!

### ISSUE THREAD

**Issue by jvino** [influxdata/telegraf#7567]
Description:

https://github.com/influxdata/telegraf/blob/release-1.14/plugins/inputs/zookeeper/zookeeper.go#L155

Why float values are not supported in this plugin?

Average values of course are float and are casted in string.
This block any will to use them as number.

Please fix it

**Comment by HarshitOnGitHub** [comment]
Can't find any reason why it shouldn't the conversion happen, I think it is just missing, WDYT @danielnelson ? Happy to fix it if so. :)

**Comment by danielnelson** [comment]
I guess this will be a bit trickier to handle than it first appears.  Switching to parsing everything as a float isn't ideal for integer values, and floats cannot represent as large of values as an int64 accurately due to having only 53-bits for the mantissa (for example see https://github.com/influxdata/telegraf/issues/7555).

This plugin uses a optimistic parse to pick the type, and this also causes problems since types can change depending on the size of the value.

The most reliable method for parsing is a table of fields and their expected types.  But if we are to support user defined values in the `mntr` response, this won't be possible.

Another method that might work is to look for a decimal place to determine float vs int.  This requires that floats always have a decimal place, even when they are an integer value.

> zk_avg_startup_snap_load_time 9.0
> zk_min_startup_snap_load_time 9

Final option would just be to switch everything to a float.  It is a bit more expensive to store than an integer and has the smaller significand, but matches more data in general.  To avoid data flipping between strings and floats we need another solution, perhaps whitelisted strings.

If we switch any types we also need to add a `metric_version = 2` option for this plugin to allow for clean migration without type conflicts.

related #7550

**Comment by jvino** [comment]
@danielnelson Thank you for you description.
I would like to ask you if the float cast could be only if the integer cast goes wrong and before the string cast. Something like this (sorry I do not use golang):
```
iVal, err := strconv.ParseInt(sValue, 10, 64)
			if err == nil {
				fields[measurement] = iVal
			} else {
                                fVal, err := strconv.ParseFloat(sValue)
                                if err == nil {
				             fields[measurement] = fVal
			        } else {
				             fields[measurement] = sValue
			}}
```
What do you think?
Thank you

**Comment by HarshitOnGitHub** [comment]
Yeah I was thinking on the same lines as @jvino mentioned. We use the same logic on redis plugin as well. But I understand what you are mentioning @danielnelson and I think redis plugin is sort of suffering from the same problem you mentioned. https://github.com/influxdata/telegraf/issues/7375
I think it is a common pattern and worth discussing. :)

**Comment by danielnelson** [comment]
We want to make sure that the selected data type is always consistent.  If we use a fallback method, which we do use in other plugins, we run into the issue where the type can change depending on it's value.  Several of our plugins using this method have open issues for this problem.

I think what would be best is to have a hardcoded type mapping for the standard fields.  My zookeeper install is a bit old, perhaps we can get an updated list, but I have:
```
zk_version      3.4.9-3--1, built on Fri, 24 May 2019 08:57:53 +0100
zk_avg_latency  0
zk_max_latency  0
zk_min_latency  0
zk_packets_received     2
zk_packets_sent 1
zk_num_alive_connections        1
zk_outstanding_requests 0
zk_server_state follower
zk_znode_count  144
zk_watch_count  0
zk_ephemerals_count     0
zk_approximate_data_size        11100
zk_open_file_descriptor_count   49
zk_max_file_descriptor_count    4096
```

Any additional fields that are not in the list should be ignored by default, but we can also add a way to include them.  There are several ways this could be done:

List of globs per field type:
```
float_include = ["*"]
string_include = []
integer_include = ["*_latency"]
unsigned_include = []
tag_include = []
```

Mapping of field patterns to type, in order of precedence:
```
extra_fields = [
  "*_latency integer",
  "* float",
]
```

**Comment by ktibi** [comment]
Hello,

Same issue here : 
`2020-08-10T08:43:20Z E! [inputs.zookeeper] Error in plugin: unexpected line in mntr response: "zk_avg_kafka-manager_read_per_namespace\t97.125"`

Any workarround ?

**Comment by wooz86** [comment]
Same issue here.

**Comment by ktibi** [comment]
Hello @danielnelson ,

did you had the time to work on this ?

I can't monitor zookeeper with telegraf right now :/

**Comment by danielmotaleite** [comment]
One more, since we upgraded zookeeper to 3.6 (from 3.5) we lost zookeeper metrics) :(

**Comment by sjwang90** [comment]
Fixed in #7533

**Comment by wooz86** [comment]
This is not fixed in #7533 @sjwang90. The values are still not converted from string to float.

**Comment by piotrp** [comment]
Some more information: all `zk_avg*_latency` metrics were changed to doubles in ZooKeeper 3.6: https://github.com/apache/zookeeper/commit/bc992480ec938a3fad4b90f75a52dd186e1b968a ([ZOOKEEPER-2641](https://issues.apache.org/jira/browse/ZOOKEEPER-2641))
